include/arch/arm/arch/32/mode/object/structures.bf

```diff
@@ -199,20 +200,21 @@ block VMFault {
 
 #ifdef CONFIG_ARM_HYPERVISOR_SUPPORT
 block VGICMaintenance {
-    field idx        6
-    field idxValid   1
-    padding         25
+    field eisr1     32
+    field eisr0     32
     padding         28
     field seL4_FaultType  4
 }
```

Comment:
Increasing the fault storage size on all architectures may be a bigger change than we want, especially because it's unrealistic for > 32 list registers to change within the same maintenance interrupt isn't it?

@lsf37, any ideas about what the impact of this change would be on verification, and then any ideas for alternative approaches? (Such as storing additional fault info with the faulting vcpu instead of in the faulting TCB?

Reply:
The verification impact of this is pretty high compared to the size of the code change, because it contains a loop, affects multiple architectures, changes type sizes etc. I'd definitely welcome exploring alternatives and also some data on actual performance difference before we merge it.

It looks to me like this actually is an API change, so this would require a (small) RFC.

Reply:
Additional considerations: what is the WCET impact, does current user space need to change, and how much effort is that going to be for people

Reply:
@lsf37 https://github.com/seL4/seL4_projects_libs/pull/105 This would be the change in the user-space

Reply:
When simple things become complex:
- (i) Increased verification efforts of the loop
- (ii) Breaking interfaces with existing VMMs
- (iii) Increased ipc msg size.

The current implementation of maintenance IRQ has inefective code:
- Reading an LR, if we already know that the LR is EOIed by reading EISR
- e.g. cleaning up pending and active IRQs - This should never happen.

Understandably nearly all parts of this MR can be done later, with exception of the first commit, if you guys are happy with it. @kent-mcleod @lsf37 @axel-h @Indanz What about the first commit for now, and later the RFC?

Reply:
> The current implementation of maintenance IRQ has inefective code:
> - Reading an LR, if we already know that the LR is EOIed by reading EISR
> - e.g. cleaning up pending and active IRQs - This should never happen.
>
> Understandably nearly all parts of this MR can be done later, with exception of the first commit, if you guys are happy with it.

I agree that the first commit improves the code quality, but I don't think it's fixing any actual errors in the existing implementation.  Sometimes the code in C ends up less optimal in order to make the verification work easier as it can be very hard to show in the proof context that certain code paths are ineffective and it's easier to add them to the C code anyway which may have been the case here (but it may also just be bad code ;) ).

For the code in the first commit, accepting the change will require additional verification work to update proofs and it would depend on the verification team's capacity for when this would get completed :
- https://github.com/seL4/l4v/blob/master/spec/haskell/src/SEL4/Object/VCPU/ARM.lhs#L528-L533 is the equivalent part in the executable spec,
- https://github.com/seL4/l4v/blob/master/proof/crefine/ARM_HYP/Syscall_C.thy#L1797C1-L2072 should be the proof that the C code corresponds to the executable spec.

Another example of ineffective code where removal requires verification effort is a condition that will always return true at the end of VGICMaintenance:`if (isRunnable(NODE_STATE(ksCurThread)))`, but removing the check is blocked on verification work to establish in the proofs that it is in fact safe to remove.

Reply:
> e.g. cleaning up pending and active IRQs - This should never happen.

I dont understand:
(ii) how can we **test** code path that never run
(ii) how can we **verify** code path that never run
(iii) What is the **worst case** that can happen if we run it?

The current sel4 vgic model doesnt remove active or pending vIRQs from the LRs.  If that happens there is an issue in the guest.

How does that help in the verification path? Runnable code is understandable, But this code never runs. 

Dead code is not good for security.